language-design.md

```diff
@@ -38,16 +128,50 @@ you know a poorly designed language? What are the symptoms?
+* allows users to expand the language in a way that feels like the underlying language (e.g., supports operator overloading).
+
+##### Programs in a well-designed language:
 
+* "[read] like prose" and are easy to understand--preferably, even by those who are not familiar with the language [Bloch, 2006].
+* has components that can be altered in a predictable way to change the meaning of the program.
+* behave in a predictable way and are obviously correct.
+
+##### A poorly-designed language:
+
+* employs trivial and unnecessary syntax.
+* places a high conceptual burden on the user.
+* requires reading lots of documentation to become familiar with the language.
+
+##### Programs in a poorly-designed language:
+* require excessive error handling.
+* have to work around the limits of the language to accomplish anything of value.
+* extend the language in an unnatural or distinct way.
+* are neither obviously correct or incorrect.
 
 ---
  
+**Question**
 
 In what way is an API a language? 
 
 **Response**
 
+People use languages to express ideas. Programmers use programming languages to express ideas to a computer. Users program with APIs to express ideas to a system, which suggests that APIs _are_ a programming language. We define languages to have both vocabulary and grammar. This holds true for programming languages as well in the form of objects and methods forming vocabulary while syntax forms a grammar. APIs provide a vocabulary by a new set of methods or objects for the user to interact with the system. Some APIs have their own grammar, while other borrow grammar from the base programming language. Either way, the API has a grammar and vocabulary, just like a language, (sort-of, more on this later). We do not know the entire domain of APIs and are unsure if there are APIs that do not have a grammar, and would not be considered a language. We could not come up with an example of an API that provides either a grammar or vocabulary but not the other. However, we do assert that languages must have both in order to be understood. Some APIs are distinct from languages in that languages are able to communicate and express ideas without an outside grammar. That is, a language does not need to borrow a grammar from another language. In that way, where an API is reliant on the grammar of the underlying programming language and does not apply new grammar rules, APIs are not languages. APIs and languages also share design principles. The themes in _Growing a Language_ and _How to Design a Good API and Why It Matters_ coincide in the goals that the authors have in mind for APIs and languages. Languages and APIs should both be designed to make expressions of ideas easier. We believe that many of the language design philosphies we ran into in class can be applied to APIs to produce an effective and natural API. We have both run into APIs that are clunky and difficult to use and could use a language-oriented design approach.
```

Comment:
> What are you actually trying to say here? What is an outside grammar? What does borrowing a grammar from another language mean?

I guess this is trying to draw a line between APIs and internal DSLs. However, the lines certainly get blurry when an API is very specific about how it can be used. To add to that, I think internal DSLs can still borrow grammar. They just might restrict, repurpose or add a little to what's there.